video: emul: virtual driver for an imager and RX engine

josuah opened the pull request:
This takes inspiration on the other skeleton.c drivers through Zephyr tree:
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/usb/udc/udc_skeleton.c
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/display/display_dummy.c

It acts as a starting point and as a puppet to explain API changes affecting image sensors.

Be used for providing test cases for the entire API, as [required for stable APIs](https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#stable).

Utilities are introduced in the process to reduce boilerplate.

Twister command (new with this PR):
```
west twister --inline-logs --platform native_sim --scenario tests/drivers/video/api/drivers.video.api
```

Fixes: #73867

josuah commented:
I note that the frame interval API was merged, this will need to include it.

- https://github.com/zephyrproject-rtos/zephyr/pull/72254

josuah commented:
- Aims to cover the most frequent use-cases and facilitate modifications on top
- Assume that image sensor drivers are apply *modes* described by pixelformat+resolution+framerate
- For every new sensor, require filling a few tables and let the ready-made functions hook them to the Zephyr video API

uLipe commented:
I like the idea of the template, my point is just from the ci perspective, it would not be good to provide something that builds? Thinking for newcomers in the contribution, they would prefer to use a template that actually builds to compare against. 

Otherwise they potentially move to copying for other sensors (which was my case when porting the gc2145).

uLipe reviewed (commented):
Just initial comments

josuah commented on drivers/video/video_sensor_skeleton.c:
```diff
@@ -0,0 +1,439 @@
+#define SENSORNAME_REG7 0x0007
+#define SENSORNAME_REG8 0x0008
+
+#define SENSORNAME_SENSOR_ID 0x99
```
For IMX219, it is `0x0219`. Hardly any variation.
For GC2145 which you worked with, it is [`0x2155`](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/video/gc2145.c#L30-L31) apparently!

Should it be an opt-in (make the config just a `#define` and let the user add the devicetree reference) or an opt-in (make the devicetree integration in `config->sensor_id` and let the user turn that in just a config)?

josuah commented on drivers/watchdog/wdt_wwdg_stm32.c:
```diff
@@ -13,6 +13,7 @@
 #include <stm32_ll_system.h>
 #include <errno.h>
 #include <zephyr/sys/__assert.h>
+#include <zephyr/sys/util.h>
```
I thought I'd replace all files which used a similar macro (separate commit) with that, but this file does not seems to make any use of the macro, and the name does not clash. I might as well remove it from the PR.

josuah commented on include/zephyr/sys/util.h:
```diff
@@ -391,6 +391,36 @@ extern "C" {
+ */
+#define ABS_DIFF(a, b) ((a) > (b) ? (a) - (b) : (b) - (a))
+#endif
+
```
It was just for this edge case:
```
printf("0x%x\n", 8U - 9);
0xffffffff
```
`(echo '#include <stdio.h>'; echo 'int main(void) { printf("0x%x\n", 8U - 9); return 0; }') | cc -xc - && ./a.out`

josuah replied:
Oh my, introducing `ABS()` is going to break everything. Currently breaking TI's HAL.
Let's not introduce this on an innocent unrelated PR yet... I only needed `ABS_DIFF()`.

josuah commented:
> Would not be good to provide something that builds? Otherwise they potentially move to copying for other sensors.

This is a good point. It is possible to replace all I2C calls by some `LOG_DBG()` and make it a *virtual* sensor, which can also be used in ztest.

See https://github.com/zephyrproject-rtos/zephyr/pull/79458#issuecomment-2399064913

> Then, the video sw generator (should be renamed though) could also go into devicetree if you decouple it into sub-devices to emulate a complex video pipeline, may be a **sensor** and an ISP-like device to just de-bayer the images come from the sensor (this way, Bayer formats can be also added in the sensor part).

josuah commented:
force-push: fix a bug when actually trying to compile it... It shows how important it is to provide code that actually compile, and why not as well: actually run.

josuah committed:
drivers: video: add emulated Imager driver and RX driver

Add a new implementation of a test pattern generator, with the same
architecture as real drivers: split receiver core and
I2C-controlled sub-device, with changes of video format in
"zephyr,emul-imager" leads to different data produced by
"zephyr,emul-rx".

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
drivers: video: common: Add utilities to seek frmival/caps structures

Introduce a video_get_format_index() utility to help finding a caps
entry out of a given format. Introduce several utilities to seek and
apply frame intervals.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
Here is a new WIP approach. Instead of having a skeleton driver that might not compile, and past this might still have bugs, here is a pair of virtual drivers:

- MIPI RX which emulates a MIPI RX core.
- Imager which emulates an image sensor

There is a `memcpy()` call to pass data from the Imager to the MIPI RX other than that, everything is implemented like it would on real hardware, example of complete implementation of existing and proposed APIs (i.e. in PRs).

Next, I need to add ztests, and see if that permits to run the samples on `native_sim`.

josuah commented:
force-push: add unit tests for the generic `video_common.h` utilities except `video_closest_frmival()`, taking a `struct video_device *` as argument.

josuah commented:
force-push: unit tests for the emulated drivers: this covers the entire video API except `set_signal()`

josuah commented:
@ngphibang you are right, I should work on a separate `wip` branch and force-push there instead of directly to the PR branch. Thank you for letting me know. Yes this is ready. Twister, check_compliance, and doc build worked locally.

josuah commented:
force-push:
- Fix 2 commits being mixed-up
- Fix bindings and deviceetree (include `i2c-device.yaml` or `base.yaml`)
- Use devicetree macros API like in https://github.com/zephyrproject-rtos/zephyr/pull/80649

josuah commented:
force-push:
- Doc build fix
- Increment the minor API version [as documented](https://docs.zephyrproject.org/latest/develop/api/overview.html#api-overview). I just found out about this page! This might need an entry in https://github.com/zephyrproject-rtos/zephyr/tree/main/doc/releases for `v4.1`

josuah commented:
In the process of implementing these emulated drivers, I introduced more video API helpers to reduce boilerplate.

If introducing helpers does not make sense in this same PR, I can integrate them back into the emulated drivers to reduce the amount of things this introduces.

ngphibang commented:
I will try to review this whenever possible, thanks !

pdgendt commented on drivers/video/video_emul_imager.c:
```diff
@@ -0,0 +1,497 @@
+	return emul_imager_write_reg(dev, EMUL_IMAGER_REG_CTRL, 0);
+}
+
+static const struct video_driver_api emul_imager_driver_api = {
```
Please use the `DEVICE_API` wrapper.

```suggestion
static DEVICE_API(video, emul_imager_driver_api) = {
```

josuah replied:
Yes thank you, done.
This will be helpful for checking device types in the video shell based on this PR!

pdgendt commented on drivers/video/video_emul_rx.c:
```diff
@@ -0,0 +1,327 @@
+	return 0;
+}
+
+static const struct video_driver_api emul_rx_driver_api = {
```
```suggestion
static DEVICE_API(video, emul_rx_driver_api) = {
```

josuah replied:
Done as well, thank you for the reminder.

loicpoulain commented on include/zephyr/drivers/video.h:
```diff
@@ -753,6 +753,60 @@ struct video_buffer *video_buffer_alloc(size_t size);
+ *
+ * @return The frame interval value in microseconds.
+ */
+uint64_t video_frmival_nsec(const struct video_frmival *frmival);
```
Could this be a simple inline helper?

josuah replied:
Yes I agree. Done now.

josuah commented:
force-push:
- fix the RGB565 pattern generator (wrong color order)
- `cfg` seems more frequently used than `conf`
- endpoint@0 (IN) first, endpoint@1 (OUT) after, in devicetree, to have the source **up**stream and sinks **down**stream in the dts files, which also matches the `media-ctl --print-dot` numbering

![scrot_20241202_081623_1271x138](https://github.com/user-attachments/assets/6030e81d-66ca-4a7f-b8ee-fc0362980c77)

loicpoulain reviewed (approved):
LGTM

ngphibang commented on include/zephyr/drivers/video.h:
```diff
@@ -753,6 +753,63 @@ struct video_buffer *video_buffer_alloc(size_t size);
+ * @return 0 when a format is found.
+ * @return -ENOENT when no matching format is found.
+ */
+int video_format_caps_index(const struct video_format_cap *fmts, const struct video_format *fmt,
```
All the introduced functions here are supposed for drivers uses only ?. I think it should be put in an internal header rather in the public header `video.h`.

josuah replied:
The users can access the format cap structure, so they could be interested in verifying: "for this particular width/height/pixelformat, what is the min/max/step?

Not sure this would ever get used that way though! Glad with putting it on a `drivers/video/video_common.h` header like often done for driver helper functions.

ngphibang commented on drivers/video/video_emul_rx.c:
```diff
@@ -0,0 +1,333 @@
+	return 0;
+}
+
+/* See #80649 */
```
As this was merged first, I will fix the conflict in https://github.com/zephyrproject-rtos/zephyr/pull/80649 then.

ngphibang replied:
s/the better way/the correct way

josuah replied:
This was opened 2 weeks before #80649, so got started with putting some custom macros, and then forgot to rebase it on #80649. Thank you for fixing these macros on #80649!

josuah commented:
I had them on `video_common.h` at first, but noticed they could also be used from the application for some purpose... so I moved them to `video.h`.

Let me know which one would be a better fit in `video_common.h`.

I also wanted to introduce IMX219 and others following this template.
So reviews will help even post-merge!
